### PR TITLE
[IOTDB-1773] Aligned timeseries support group by query without value filter for standalone mode

### PR DESCRIPTION
**Pull request by liuminghui233**

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithoutValueFilterDataSet.java [hunk 1]
```diff
@@ -56,22 +55,13 @@
 
   protected Map<PartialPath, GroupByExecutor> pathExecutors = new HashMap<>();
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithoutValueFilterDataSet.java:56]: Since you have have already split all the process of aligned and non-align path, you can replace the `pathExecutors ` to two `Map` like `resultIndexes` and `alignedPathToIndexesMap`. If so, you don't have to cast them to corresponding exact class each time.

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+
+    // read chunk data secondly
+    if (readAndCalcFromChunk(curStartTime, curEndTime)) {
+      return results;
+    }
+
+    // read from file
+    while (reader.hasNextFile()) {
+      // try to calc from fileMetaData
+      Statistics fileStatistics = reader.currentFileStatistics();
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java:158]: `reader.currentFileStatistics()` return the first sub sensor's statistics. Here you need to get the time column's statistics which could never be null, so you don't even to do the `isEmpty` judgement in the next line.

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+      if (fileStatistics != null && !isEmpty(fileStatistics)) {
+        if (fileStatistics.getStartTime() >= curEndTime) {
+          return results;
+        }
+        if (reader.canUseCurrentFileStatistics()
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java:163]: The `canUseCurrentFileStatistics()` function in `AlignedSeriesAggregateReader` has the same wrong thing like the above, it use the first sub sensor's statistics instead of time column's. You can change that by the way.

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+    }
+
+    // read from file
+    while (reader.hasNextFile()) {
+      // try to calc from fileMetaData
+      Statistics fileStatistics = reader.currentFileStatistics();
+      if (fileStatistics != null && !isEmpty(fileStatistics)) {
+        if (fileStatistics.getStartTime() >= curEndTime) {
+          return results;
+        }
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java:162]: You still need to use `ascending` here, like the judgement in `readAndCalcFromChunk` function. It may have bugs even in non-aligned case, you can make another pr to fix that before this pr merging

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java [hunk 5]
```diff
@@ -0,0 +1,388 @@
...
+      }
+      result.updateResultFromStatistics(statistics);
+    }
+  }
+
+  private boolean readAndCalcFromChunk(long curStartTime, long curEndTime)
+      throws IOException, QueryProcessException {
+    while (reader.hasNextChunk()) {
+      // try to calc from chunkMetaData
+      Statistics chunkStatistics = reader.currentChunkStatistics();
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java:202]: same as above ,use time column's statistics.

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java [hunk 6]
```diff
@@ -0,0 +1,388 @@
...
+      if (chunkStatistics != null && !isEmpty(chunkStatistics)) {
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java:203]: no need to judge whether it is empty.

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java [hunk 7]
```diff
@@ -0,0 +1,388 @@
...
+        if (chunkStatistics.getStartTime() >= curEndTime) {
+          if (ascending) {
+            return true;
+          } else {
+            reader.skipCurrentChunk();
+            continue;
+          }
+        }
+        if (reader.canUseCurrentChunkStatistics()
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java:212]: change `canUseCurrentChunkStatistics` function to use time column in another pr.

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java [hunk 8]
```diff
@@ -0,0 +1,388 @@
...
+    return false;
+  }
+
+  private boolean readAndCalcFromPage(long curStartTime, long curEndTime)
+      throws IOException, QueryProcessException {
+    while (reader.hasNextPage()) {
+      // try to calc from pageHeader
+      Statistics pageStatistics = reader.currentPageStatistics();
+      // must be non overlapped page
+      if (pageStatistics != null && !isEmpty(pageStatistics)) {
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java:238]: same as above

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java [hunk 9]
```diff
@@ -0,0 +1,388 @@
...
+        // current page max than time range
+        if (pageStatistics.getStartTime() >= curEndTime) {
+          if (ascending) {
+            return true;
+          } else {
+            reader.skipCurrentPage();
+            continue;
+          }
+        }
+        if (reader.canUseCurrentPageStatistics()
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java:248]: change it in another pr.

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java [hunk 10]
```diff
@@ -0,0 +1,388 @@
...
+      if (batchData == null || !batchData.hasCurrent()) {
+        continue;
+      }
+
+      // stop calc and cached current batchData
+      if (ascending && batchData.currentTime() >= curEndTime) {
+        preCachedData = batchData;
+        // reset the last position to current Index
+        lastReadCurArrayIndex = batchData.getReadCurArrayIndex();
+        lastReadCurListIndex = batchData.getReadCurListIndex();
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java:275]: put this before the if block and delete the following same two lines, just like `LocalGroupByExecutor`

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java [hunk 11]
```diff
@@ -0,0 +1,388 @@
...
+    }
+    return true;
+  }
+
+  private boolean isEmpty(Statistics statistics) {
+    if (statistics.getStartTime() == Long.MAX_VALUE && statistics.getEndTime() == Long.MIN_VALUE) {
+      return true;
+    }
+    return false;
+  }
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java:387]: delete it

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalGroupByExecutor.java [hunk 12]
```diff
@@ -71,18 +68,19 @@ public LocalGroupByExecutor(
...
     timeFilter = queryDataSource.updateFilterUsingTTL(timeFilter);
-    this.reader =
+    // init SeriesAggregateReader for non-aligned series
+    reader =
         new SeriesAggregateReader(
             path,
-            allSensors,
-            dataType,
+            Collections.singleton(path.getMeasurement()),
+            path.getSeriesType(),
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalGroupByExecutor.java:76]: Why change this? Here is a query improvement before.

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalGroupByExecutor.java [hunk 13]
```diff
@@ -150,12 +148,6 @@ private void calcFromBatch(BatchData batchData, long curStartTime, long curEndTi
         result.updateResultFromPageData(batchIterator, curStartTime, curEndTime);
       }
     }
-    lastReadCurArrayIndex = batchData.getReadCurArrayIndex();
-    lastReadCurListIndex = batchData.getReadCurListIndex();
-    // can calc for next interval
-    if (batchData.hasCurrent()) {
-      preCachedData = batchData;
-    }
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalGroupByExecutor.java:158]: You delete these lines in `calcFromBatch` and then add them in the process of `readAndCalcFromPage`, but `calcFromBatch` is still called by `calcFromCacheData` where you forgot to change. So why you want move these lines outside of this method.

File: server/src/main/java/org/apache/iotdb/db/query/reader/series/AlignedSeriesAggregateReader.java [hunk 14]
```diff
@@ -174,4 +174,8 @@ public int getCurIndex() {
   public void resetIndex() {
     curIndex = 0;
   }
+
+  public int getSubSensorSize() {
+    return subSensorSize;
+  }
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/query/reader/series/AlignedSeriesAggregateReader.java:180]: This method is not used, remove it.

**Review by JackieTien97 — CHANGES_REQUESTED** [review]
Add more cases in IT, like `select count(*) from root.sg1.d1`, select count(*) from root.sg1.* to test query with `*`. And You should also add some tests while there are deletion. You can refer to IoTDBRawQueryWithoutValueFilterIT, IoTDBRawQueryWithoutValueFilterWithDeletionIT, IoTDBRawQueryWithoutValueFilter2IT and IoTDBRawQueryWithoutValueFilterWithDeletion2IT.

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java (hunk 4 above)
Comment (liuminghui233) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java:162]: fix in #4448

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java (hunk 7 above)
Comment (liuminghui233) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java:212]: fix in #4451

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java (hunk 9 above)
Comment (liuminghui233) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java:248]: fix in #4451

File: cluster/src/main/java/org/apache/iotdb/cluster/query/groupby/ClusterGroupByFillNoVFilterDataSet.java [hunk 15]
```diff
@@ -42,7 +42,6 @@ public ClusterGroupByFillNoVFilterDataSet(
     this.metaGroupMember = metaGroupMember;
   }
 
-  @Override
```
[the remark this answers is not in the document]
Comment (liuminghui233) [reply cluster/src/main/java/org/apache/iotdb/cluster/query/groupby/ClusterGroupByFillNoVFilterDataSet.java:45]: fixed

File: cluster/src/main/java/org/apache/iotdb/cluster/query/groupby/ClusterGroupByNoVFilterDataSet.java [hunk 16]
```diff
@@ -48,7 +48,6 @@ public ClusterGroupByNoVFilterDataSet(
     this.metaGroupMember = metaGroupMember;
   }
 
-  @Override
```
[the remark this answers is not in the document]
Comment (liuminghui233) [reply cluster/src/main/java/org/apache/iotdb/cluster/query/groupby/ClusterGroupByNoVFilterDataSet.java:51]: fixed

File: server/src/main/java/org/apache/iotdb/db/metadata/utils/MetaUtils.java [hunk 17]
```diff
@@ -146,4 +148,52 @@ public static void collectLastNode(IMNode node, List<IMNode> lastNodeList) {
...
+          // groupPath is changed here so we update it
+          List<List<Integer>> subIndexes = result.remove(groupPath);
+          subIndexes.add(indexes);
+          groupPath.addMeasurement((MeasurementPath) seriesPath);
+          result.put(groupPath, subIndexes);
+        }
+      }
+    }
+    return result;
+  }
```
[the remark this answers is not in the document]
Comment (liuminghui233) [reply server/src/main/java/org/apache/iotdb/db/metadata/utils/MetaUtils.java:198]: fixed

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithoutValueFilterDataSet.java (hunk 1 above)
Comment (liuminghui233) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithoutValueFilterDataSet.java:57]: fixed

File: server/src/main/java/org/apache/iotdb/db/query/reader/series/AlignedSeriesAggregateReader.java (hunk 14 above)
Comment (liuminghui233) [reply server/src/main/java/org/apache/iotdb/db/query/reader/series/AlignedSeriesAggregateReader.java:180]: fixed

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithoutValueFilterDataSet.java [hunk 18]
```diff
@@ -56,22 +55,13 @@
...
 
-  /**
-   * path -> result index for each aggregation
-   *
-   * <p>e.g.,
-   *
-   * <p>deduplicated paths : s1, s2, s1 deduplicated aggregations : count, count, sum
-   *
-   * <p>s1 -> 0, 2 s2 -> 1
-   */
```
[the remark this answers is not in the document]
Comment (liuminghui233) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithoutValueFilterDataSet.java]: fixed

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalGroupByExecutor.java (hunk 13 above)
Comment (liuminghui233) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalGroupByExecutor.java:158]: fixed

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java (hunk 11 above)
Comment (liuminghui233) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java:368]: fixed

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java (hunk 10 above)
Comment (liuminghui233) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java:256]: fixed

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java (hunk 2 above)
Comment (liuminghui233) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java:128]: fixed

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java (hunk 3 above)
Comment (liuminghui233) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java:133]: fixed

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java (hunk 5 above)
Comment (liuminghui233) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java:183]: fixed

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java (hunk 6 above)
Comment (liuminghui233) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java:184]: fixed

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java (hunk 8 above)
Comment (liuminghui233) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java:219]: fixed
